documentation/proposals/Proposal - Generic Math.md

```diff
@@ -0,0 +1,366 @@
+  - A Static implementation of these functions **must** be available
+  - Define matching `*` operators which function the same
+- Define TransformNormal functions which take a Matrix of higher dimensionality (Vector 2 can use Matrix2xn, Matrix3xn, and matrix4xn) and return a vector containing the output (type should match the outer type e.g. Vector2.Transform(Matrix4x4) returns Vector2)
+  - A Static implementation of these functions **must** be available
+
+For I types, the following requirements **must** also be fulfilled:
+- the bitwise `&`, `|`, and `^` operators defined between two vectors which returns a vector which has had these operators applied on a component-wise basis.
+- the bitwise `&`, `|`, and `^` operators defined between a vectors and a scalar value that matches the generic type which returns a vector which has had these operators applied on a component-wise basis with the scalar.
+- the unary bitwise `~` operator defined which negates the bits of the vector components.
+- AllBitsSet property which returns a Vector with all bits set to 1
+- Define the following functions to match available IBinaryInteger (Vector replaced with type, e.g. `Vector2I<T>`):
+  - GetByteCount()
+    - Returns an int representing the number of bytes that will be written as part of TryWriteLittleEndian(Span<byte>, out int).
+  - GetShortestBitLength()
+    - Returns an int representing the length, in bits, of the shortest two's complement representation of the current value.
+  - TryWriteBigEndian(Span<byte> destination, out in bytesWritten)
+    - Returns boolean of whether the operation was successful
+  - WriteBigEndian(Span<byte> destination)
+    - Returns int representing the bytes written
+  - WriteBigEndian(Span<byte> destination, int startIndex)
+    - Returns int representing the bytes written
+  - WriteBigEndian(byte[] destination)
+    - Returns int representing the bytes written
+  - TryWriteLittleEndian(Span<byte> destination, out in bytesWritten)
+    - Returns boolean of whether the operation was successful
+  - WriteLittleEndian(Span<byte> destination)
+    - Returns int representing the bytes written
+  - WriteLittleEndian(Span<byte> destination, int startIndex)
+    - Returns int representing the bytes written
+  - WriteLittleEndian(byte[] destination)
+    - Returns int representing the bytes written
+- Define the following static functions for these types to match IBinaryInteger (Vector replaced with type, e.g. `Vector2I<T>`) which returns a new vector with these operations applied component-wise, unless otherwise specified:
+  - Log2(Vector x)
+  - DivRem(Vector left, Vector right)
+    - Returns tuple of 2 Vectors (Vector Quotient, Vector Remainder)
+  - PopCount(Vector x)
+    - returns 
```

Comment:
returns what?

Reply:
whoops I must have got distracted writting this, fixed